Xl python inference

lopez-hector opened the pull request:
1. Added XL inference capabilities to python script.
    1. Infers XL from model version (could change to flag to mimic swift behavior)
    2. Select dtype for VAE decoder from `expected_inputs`
    3. Schedulers return either np.ndarray or torch.Tensor so i put some logic to catch this and convert to numpy array
    4. Added `_get_add_time_ids()`

Retains backwards compatibility. 

"A high quality photo of a surfing dog" 
num_inference_steps = 25
seed = 93
guidance_scale = 10

<img width="959" alt="image" src="https://github.com/apple/ml-stable-diffusion/assets/65481379/f53aaf75-ca34-41f7-85bd-9e183366f0be">


Thank you for your interest in contributing to Core ML Stable Diffusion! Please review [CONTRIBUTING.md](../CONTRIBUTING.md) first. If you would like to proceed with making a pull request, please indicate your agreement to the terms outlined in CONTRIBUTING.md by checking the box below. If not, please go ahead and fork this repo and make your updates.

We appreciate your interest in the project!

Do not erase the below when submitting your pull request:
#########

- [ X] I agree to the terms outlined in CONTRIBUTING.md

atiorh commented:
Thanks @lopez-hector! Are you using the main branch to export SDXL models? I ran a quick test and my locally exported models failed with:
```
  File "ml-stable-diffusion/python_coreml_stable_diffusion/pipeline.py", line 132, in _encode_prompt
    text_embeddings = self.text_encoder(
KeyError: 'last_hidden_state'
```

atiorh commented:
Thanks @lopez-hector! I ran a quick test and received the following error:
```
File "ml-stable-diffusion/python_coreml_stable_diffusion/coreml_model.py", line 71, in _verify_inputs
    raise TypeError(
TypeError: Expected shape (2, 6), got (12,) for input: time_ids
```

lopez-hector commented:
Interesting, I don't get this issue when using the converted packages from https://huggingface.co/apple/coreml-stable-diffusion-xl-base/tree/main/packages. I just redownloaded them and tested.

Using these packages from hugging face the expected inputs for the UNET are:

```
{'encoder_hidden_states': {'dtype': <class 'numpy.float16'>,
                           'shape': (2, 2048, 1, 77)},
 'sample': {'dtype': <class 'numpy.float16'>, 'shape': (2, 4, 128, 128)},
 'text_embeds': {'dtype': <class 'numpy.float16'>, 'shape': (2, 1280)},
 'time_ids': {'dtype': <class 'numpy.float16'>, 'shape': (12,)},
 'timestep': {'dtype': <class 'numpy.float16'>, 'shape': (2,)}}
```

Im not sure how those were converted before being uploaded to HF. What are the expected inputs for your converted packages?

atiorh commented:
I see! I am currently testing models exported from #227 and the different input shapes are handled on the [Swift](https://github.com/apple/ml-stable-diffusion/pull/227/files#diff-c41606fda12b1aa1e46d0f90149654c065c0c0618bc822c5b1c21210e52515afR152) code path but not on the Python code path. cc: @ZachNagengast

atiorh commented:
@lopez-hector Do you mind adding a similar support for accepting both input shapes on the Python code path?

atiorh commented:
@lopez-hector There is one more thing I was thinking about improving about the Python inference code path which is to switch to [coremltools.models.CompiledMLModel](https://github.com/apple/coremltools/blob/7.0/coremltools/models/_compiled_model.py#L18) from `coremltools.models.MLModel` so that the results from model compilation (which bloats the loading time from seconds to minutes) are cached during first load and not recompiled after each load. If you would like, we can fold those changes into this PR or I can make a separate one. Please let me know.

lopez-hector commented:
Will add support for input shapes soon.

Happy to include the change for CompileMLModel here. I was hoping there was a way to speed that up!

atiorh commented:
@lopez-hector  Thanks!

atiorh commented:
Note that the `CompiledMLModel` will need to load the `.mlmodelc` assets generated via `--bundle-resources-for-swift-cli` instead of the original `.mlpackage` assets so the different filename conventions will need to be handled, just a heads-up

lopez-hector commented:
> I see! I am currently testing models exported from #227 and the different input shapes are handled on the [Swift](https://github.com/apple/ml-stable-diffusion/pull/227/files#diff-c41606fda12b1aa1e46d0f90149654c065c0c0618bc822c5b1c21210e52515afR152) code path but not on the Python code path. cc: @ZachNagengast

What XL models were giving you a shape of (2, 6). I added the shape handling and would like to run a quick test.

lopez-hector commented:
Hi @atiorh, I added support for the CompiledMLModel. On my M2 Max MacBook, it reduces loading time by half for the text encoders and VAE. About 3/4 reduction for the UNET.

UNET takes about 15s to load from compiled sources.

- added logic to detect `mlpackage` or `mlmodelc` from the input directory.
- added flag `--model-sources` to force loading from either `packages` or `compiled` sources
- If loading from .mlmodelc, we load using CompiledMLModel and generate expected_inputs from `metadata.json`

atiorh commented:
Excellent, thanks @lopez-hector!
